### PR TITLE
Use `sudo` for symlinks if necessary.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Fixes https://github.com/Homebrew/homebrew-cask/pull/147086#issuecomment-1546322878.

### CODE REVIEW

File: Library/Homebrew/cask/artifact/symlinked.rb [hunk 1]
```diff
@@ -64,19 +64,20 @@ def link(force: false, **options)
         end
 
         ohai "Linking #{self.class.english_name} '#{source.basename}' to '#{target}'"
-        create_filesystem_link(**options)
+        create_filesystem_link(command: command)
       end
 
       def unlink(**)
```
Comment (Bo98) [inline Library/Homebrew/cask/artifact/symlinked.rb:70]:
```suggestion
      def unlink(command: nil, **)
```

I think?

File: Library/Homebrew/cask/artifact/symlinked.rb [hunk 2]
```diff
@@ -64,19 +64,20 @@ def link(force: false, **options)
...
         return unless target.symlink?
 
         ohai "Unlinking #{self.class.english_name} '#{target}'"
-        target.delete
+        Utils.gain_permissions_remove(target, command: command)
```
Comment (Bo98) [inline Library/Homebrew/cask/artifact/symlinked.rb:74]: Looking inside `gain_permissions_remove`, I'm not convinced this behaves correctly - shouldn't the symlink condition come first?

Comment (reitermarkus) [reply Library/Homebrew/cask/artifact/symlinked.rb:74]: Yeah, I think we want to handle `file?` and `symlink?` separately in there.

File: Library/Homebrew/cask/artifact/symlinked.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/cask/artifact/symlinked.rb:70]: Right, thanks.

**Review by Bo98 — APPROVED** [review]

**Pull request merged**